Add ADR 0004 - Naming of Interface Implementations (#7794)
thunderbird/thunderbird-android: docs/architecture/adr/0004-naming-of-interface-implementations.md

> # Naming of Interface Implementations

Reviewer: Alternative

```suggestion
# Naming Conventions for Interfaces and Their Implementations
```

## Status

- **Proposed**

## Context

When there's an interface that has multiple implementations it's often easy enough to give meaningful names to both the
interface and the implementations (e.g. the interface `Backend` with the implementations `ImapBackend` and
`Pop3Backend`). Naming becomes harder when the interface mainly exists to allow having isolated unit tests and the
production code contains exactly one implementation of the interface.
Prior to this ADR we didn't have any naming guidelines and the names varied widely. Often when there was only one
(production) implementation, the class name used one of the prefixes `Default`, `Real`, or `K9`. None of these had any
special meaning and it wasn't clear which one to pick when creating a new interface/class pair.

## Decision

> We'll be using the following guidelines for naming interfaces and classes implementing an interface:

Reviewer: ```suggestion
We'll be using the following guidelines for naming interfaces and their implementation classes:
```

> - Name interfaces like you would name a class. Do not use the `IInterface` pattern. Try to keep the name short.
> - Name implementations by adding a prefix to the interface name, e.g. `InMemoryThing` for an interface named `Thing`.
> - Try to use a descriptive name whenever possible, e.g. `DatabaseMessageStore` instead of `DefaultMessageStore`.
> - When adding an interface to isolate platform specific behavior, use the platform name as the prefix, e.g.
>   `AndroidPowerManager`.
> - When there are app-specific implementations of an interface, use the prefix `K9` for K-9 Mail and `Tb` for
>   Thunderbird. Example: `K9AppNameProvider` and `TbAppNameProvider`.
> - When there's no (short) descriptive name and there's only one production implementation of an interface, use the
>   prefix `Default`. Example: `DefaultImapFolder`.

Reviewer: I added highlight for the sections to give a better overview and changed the descriptions.

```suggestion
1. **Interface Naming:** Name interfaces as if they were classes, using a clear and descriptive name. Avoid using the "IInterface" pattern.
2. **Implementation Naming:* Use a prefix that clearly indicates the relationship between the interface and implementation, such as `DatabaseMessageStore` or `InMemoryMessageStore` for the `MessageStore` interface.
3. **Descriptive Names:** Use descriptive names for interfaces and implementing classes that accurately reflect their purpose and functionality.
4. **Platform-specific Implementations:** Use the platform name as a prefix for interface implementations specific to that platform, e.g. `AndroidPowerManager`.
5. **App-specific Implementations:** Use the prefix`K9` for K-9 Mail and `Tb` for
  Thunderbird when app-specific implementations are needed, e.g. `K9AppNameProvider` and `TbAppNameProvider`.
6. **Flexibility:** If no brief descriptive name fits and there is only one production implementation, use the prefix `Default`, like `DefaultImapFolder`.
```

## Consequences

- **Positive Consequences**

>   - This gives us more consistent naming of interface implementations.

Reviewer: ```suggestion
- Improved code readability and maintainability through consistent naming.
- Reduced confusion and misunderstandings by using clear and descriptive names.
```

- **Negative Consequences**

>   - Some initial effort is required to rename existing classes that don't follow this naming scheme.

Reviewer: ```suggestion
  - Initial effort is required to rename existing classes that do not follow these naming conventions.
```
